### ISSUE TITLE
Truffle debugger hangs while stepping instruction

### ISSUE THREAD

**Issue by Poytr1** [ConsenSys-archive/truffle#5856]
- [x] I've [opened a support ticket](https://trufflesuite.zendesk.com/hc/en-us/requests/new) before filing this issue.

---------------------------

## Issue

Truffle debugger hangs while stepping instruction

## Steps to Reproduce

```
truffle debug 0x65d06d2c333b3705554cbb2ee19897e385787a7745939b30dc857161e488671c --fetch-external --url localhost:8545
```
Then step instruction for 5232 instructions:
```
;5232
```
Then step instruction,
![Screenshot 2023-01-26 at 00 08 34](https://user-images.githubusercontent.com/20534718/214615299-84388dc6-85c0-4f53-bf10-9490377e308a.png)
The debugger hangs.

## Expected Behavior

Step to the next instruction

## Actual Results

It never returns

## Environment

* Operating System: MacOS
* Ethereum client:
* Truffle version (`truffle version`): v5.7.0
* node version (`node --version`):v18.2.0
* npm version (`npm --version`): 8.9.0

**Comment by eggplantzzz** [comment]
Hey @Poytr1, thanks for reporting this! Could you let me know how long the debugger hangs? Does it ever return? I just ask as it is possible that it is performing some intense computation. Also, what client/setup are you using?

**Comment by eggplantzzz** [comment]
Thanks for the added information! @haltman-at

**Comment by haltman-at** [comment]
@Poytr1 One more question -- it looks like you're debugging a verified external contract on (a forked verison of) some network.  What network?  Is that mainnet?  I was just setting up to reproduce this and realized that was never specified.  I'll assume for now that it's mainnet, but please let me know.  Thanks!

**Comment by haltman-at** [comment]
Well, I reproduced the problem so I guess it is mainnet! :)  Thanks for bringing this to our attention!

**Comment by haltman-at** [comment]
OK, on looking into this, I'm pretty confused as to what's going on here.

Let's start with what I can determine.  If you attempt to advance from the highlighted step, the debugger will attempt to decode the `checkpoints` array.  (This is because of the debugger's txlog feature, which admittedly isn't very user-visible at the moment.  However the same code would be called if you tried to view the contents of `checkpoints` with `v` instead at any point.)

The array ` checkpoints` has -- at least according to what the debugger can determine -- a length of 45,040 elements.  (Each element is a struct consisting of two `uint128`s.)  Ideally there would be some way to paginate that, but, that being something we've never implemented, the debugger will attempt to decode all of it.  Most of it is storage that the debugger can't currently see, mind you, so it's going to decode to question marks.

After decoding about 900 of these -- the exact number varies -- the debugger stopds decoding and hangs.  I don't know why.  Not only does the count vary, but the exact point in the code where the hang happens seems to vary as well.  Again, I don't know why.

So I'm fairly stumped at the moment.  I don't know what's going on here.  Those are the pieces of the puzzle I can determine, however.

**Comment by haltman-at** [comment]
> BTW, I'm curious how you debugged to where the checkpoint is being decoded, are there any instructions I can follow to debug this myself using truffle code?

I mean, it's not something I would really expect a person to be able to figure out without good knowledge of the internals of Truffle Debugger.  I determined this by adding a bunch of debug logging, but without knowing how the debugger works it would have been difficult to determine where to add such logging.  I started by determining which of the debugger's 5 main submodules was hanging (it was txlog), and then I dug in further to see where the hang was occurring.  This then took me outside `debugger` proper and into `codec`, where most of the decoding occurs.  There's a lot going on in the debugger and without some foreknowledge of how it works I don't see how one would reasonably debug it.

**Comment by haltman-at** [comment]
No, we haven't decided what to do about this yet, sorry.  There is no presently-available workaround if that's what you were wondering.

**Comment by haltman-at** [comment]
OK so @gnidan and I talked about this and we think it makes sense to add a command-line option that would turn off the part that's causing the problem.  We're stretched a bit thin at the moment so it may be some more time before we get to this, sorry!  But yes I think we can do something about this.
